### PR TITLE
Upgrade to new `asm` syntax

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

closes: #305

There is one change in the assembly here. Earlier, we used to declare `P::INV` as an immediate input for the assembly block, but due to limitations of `const` in Rust, we can no longer do that, and must declare it as an input. We can probably work around this manually, but it would make things slightly uglier. I would benchmark the performance difference, but unfortunately I don't have access to an x86 machine atm. Could one of @jon-chuang or @ValarDragon benchmark this PR?

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by Pratyush** [comment]
This is going to be stabilized soon: https://github.com/rust-lang/rust/issues/72016#issuecomment-964186287

I'd like to check whether the strategy in [this playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=2dd56e74f835ad95ee9460e5c7898a2d) would allow for a cleaner impl (eg: getting rid of the funky macro for constructing the `asm` block), but otherwise we should move quickly to update to the new syntax.

<details>
<summary>Copy of playground code for posterity</summary>
```rust
#![feature(asm)]

mod intrinsics {
    #[inline(always)]
    pub unsafe fn adcx(a: &mut u64, b: u64) {
        asm! {
            "adcx {0}, {1}",
            inout(reg) *a,
            in(reg) b
        };
    }

    #[inline(always)]
    pub unsafe fn adox(a: &mut u64, b: u64) {
        asm! {
            "adox {0}, {1}",
            inout(reg) *a,
            in(reg) b
        };
    }

    #[inline(always)]
    pub unsafe fn clear_flags() {
        asm! {
            "xor {0}, {0}",
            out(reg) _
        }
    }

    // doesn't inline properly
    // pub use std::arch::x86_64::_mulx_u64 as mulx;

    #[inline(always)]
    pub unsafe fn mulx(a: u64, b: u64, lo: &mut u64, hi: &mut u64) {
        asm!(
            "mulx {2}, {1}, {0}",
            in(reg) b,
            lateout(reg) *lo,
            lateout(reg) *hi,
            in("rdx") a
        );
    }
}

/// doesn't do actual work, just demonstration
pub fn carry_chain(a: [u64; 2], b: [u64; 2]) -> u64 {
    use intrinsics::*;
    unsafe {
        // clear flags CF and OF
        intrinsics::clear_flags();
        let mut hi = 0;
        let mut lo = 0;

        mulx(a[0], b[0], &mut lo, &mut hi);
        adox(&mut lo, hi);
        adcx(&mut hi, b[0]);

        mulx(a[1], b[1], &mut lo, &mut hi);
        adox(&mut lo, hi);
        adcx(&mut hi, b[1]);

        lo
    }
}


#[test]
fn test_adcx() {
    let mut a = 5;
    let b = 3;
    unsafe {
        intrinsics::adcx(&mut a, b);
    }
    assert_eq!(a, 8);
}


#[test]
fn test_mulx() {
    let a = 5;
    let b = 3;
    let mut hi = 0;
    let mut lo = 0;
    unsafe {
        intrinsics::mulx(a, b, &mut lo, &mut hi);
    }
    assert_eq!(lo, 5 * 3);
    assert_eq!(hi, 0);
}
```
</details>

**Comment by weikengchen** [comment]
There is a decision to make:

- Since 1.59.0, `asm!` is now stable.
- However, `asm_const` is not stable, but we do use `const` in asm in certain places.

Ideas? We can still have the assembly in nightly.

**Comment by weikengchen** [comment]
Based on the CI, there might be a hard-to-find problem:

```
Error: test failed, to rerun pass '-p ark-ff --lib'
Caused by:
  process didn't exit successfully: `/home/runner/work/algebra/algebra/target/debug/deps/ark_ff-a63347d5ec55c228` (signal: 11, SIGSEGV: invalid memory reference)
```

File: ff/src/lib.rs
Comment (Pratyush) [inline ff/src/lib.rs:13]: I think we can just get rid of these? It's already stable in nightly, right?

**Review by weikengchen — APPROVED** [review]

File: ff/Cargo.toml
Comment (Pratyush) [inline ff/Cargo.toml:39]: Hm I don't think we should enable it by default, no? It's still unsafe that people might not want.

Comment (weikengchen) [reply ff/Cargo.toml:39]: then let me add it back.

Comment (weikengchen) [reply ff/Cargo.toml]: Resolved.

**Comment by weikengchen** [comment]
One thing to check is whether the code can compile in non-x86-64 machines (aka arm). In the build.rs, I use `is_x86_feature_detected!`. I worried that it would be undefined in arm.

**Comment by weikengchen** [comment]
For the previous comment---I think check_std does that job.

File: ff/build.rs
Comment (Pratyush) [inline ff/build.rs:21]: Hm should we just get rid of this, and instead change our minimum supported rust version to 1.59? And in the places that we use the assembly backend, we can directly gate it with a 
```rust
#[cfg(
    feature = "asm",
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
)]
```

Comment (weikengchen) [reply ff/build.rs:21]: I think there is only one question: do we want to bump up the minimal supported rust version to 1.59?

That would clearly be ideal, and we can cut the `build.rs`.

Comment (Pratyush) [reply ff/build.rs:21]: I think it's fine personally

Comment (weikengchen) [reply ff/build.rs:21]: I would recommend doing that in a separate PR---the current stable is 1.58.1, and it would need 20+ days to be 1.59.0.

So, if we drop it, for the next three weeks, we may have some trouble.

Comment (weikengchen) [reply ff/build.rs:21]: I can add an issue and a separate PR.

Comment (Pratyush) [reply ff/build.rs:21]: Hm then I would still move most of the `cfg` logic out of the `build.rs`, but emit the `asm` cfg in `build.rs`, so that the `cfg` gate becomes
```rust
#[cfg(all(
    asm,
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
))]
```

Comment (weikengchen) [reply ff/build.rs:21]: and the `build.rs` becomes the following? So it would in a shape that is ready to drop.

```
use rustc_version::Version;

fn main() {
    println!("cargo:rerun-if-changed=build.rs");
    
    let rust_version = rustc_version::version().unwrap();
    let is_asm_stable = rust_version >= Version::new(1, 59, 0);
    let asm_enabled = cfg!(feature = "asm");

    let should_use_asm = is_asm_stable && asm_enabled;
    if should_use_asm {
        println!("cargo:rustc-cfg=use_asm");
    }
}
```

Comment (Pratyush) [reply ff/build.rs:21]: Yup, exactly!

Comment (weikengchen) [reply ff/build.rs:21]: There is a complication: what do you think would be the best solution here?

https://github.com/arkworks-rs/algebra/blob/migrate-to-asm/ff/src/fields/models/fp/montgomery_backend.rs#L86

Comment (weikengchen) [reply ff/build.rs:21]: ah, it seems that I can simply do:

```
if N <= 6
                && N > 1
                && cfg!(all(
                    asm,
                    target_feature = "bmi2",
                    target_feature = "adx",
                    target_arch = "x86_64"
                ))
            {
```

Comment (Pratyush) [reply ff/build.rs:21]: We write `cfg!(all(
    use_asm,
    target_feature = "bmi2",
    target_feature = "adx",
    target_arch = "x86_64"
))`

Comment (weikengchen) [reply ff/build.rs:21]: another complication---actually this is a mistake:
https://github.com/arkworks-rs/algebra/blob/migrate-to-asm/ff/src/biginteger/mod.rs#L242

It just detects the `asm` feature flag but not other things. 

Seems to be complicated. 

Should we do `cfg_attr!` in the `lib.rs`?

Comment (Pratyush) [reply ff/build.rs:21]: That one is fine, as it doesn't require 1.59, nor `bmi2` or `adx`

Comment (weikengchen) [reply ff/build.rs]: Changed the logic. It appears that rust version has to be queried in build.rs. The rest can be moved out. 

I use `inline_asm_stable` since not all asm requires 1.59.0, as you pointed out.

Comment (weikengchen) [reply ff/build.rs]: `cfg(version(1.59))` is not yet stable, though, we can drop it when we decided that the minimal supported rust version is >=1.59.

File: CHANGELOG.md
Comment (Pratyush) [inline CHANGELOG.md:8]:
```suggestion
```

Comment (weikengchen) [reply CHANGELOG.md]: The other CHANGELOG entry also no longer applies. Any suggestion?

File: CHANGELOG.md
Comment (Pratyush) [inline CHANGELOG.md:42]:
```suggestion
- [\#306](https://github.com/arkworks-rs/algebra/pull/306) (`ark-ff`, `ark-ff-asm`) Make the assembly backend available on `stable`.
```

**Comment by Pratyush** [comment]
Thanks for getting this over the line, @weikengchen!

There's one remaining TODO that we should file an issue for: in the previous version of the backend, `Self::INV` was a constant in the generated assembly, but in the current version it's passed in a register. This decreases the number of available registers, increasing register and hence memory pressure.

Unfortunately making `Self::INV` into a constant requires stabilization of the `asm_const` and (part of) the `generic_const_exprs` features, so it might take a while.

**Comment by weikengchen** [comment]
Talking about which, since we have already used the ugly `movq_zero!`, would it make sense to just hardcode the constants?

There is an issue---if we hardcode that as a constant, does it mean that Rust would generate different assembly codes for each different field? Or, is that already happening?

**Comment by weikengchen** [comment]
My feeling is that `asm_const` is mostly a Rust side problem because it needs to be implemented in a general manner, but we are doing a very specific and simple case (`movq`).

**Comment by Pratyush** [comment]
We can't quite hardcode the constants, because we don't know what they are until instantiation time in some downstream crate. We *could* provide a macro that's the equivalent of https://crates.io/crates/ff_derive, which people invoke on the `MontConfig` in the appropriate downstream crate, and that macro could generate the assembly with full knowledge of the various involved constants.

**Comment by weikengchen** [comment]
...and this approach should also allow us to hardcode `modulus`, which at this moment is a memory access?

**Comment by Pratyush** [comment]
Yeah, it would allow us to hardcode modulus also.

**Comment by weikengchen** [comment]
With that, I think we are reaching the destination of single-field-element asm optimization. 

To improve further, we will need to do batching and SIMD.

**Comment by Pratyush** [comment]
@weikengchen is this ready to merge? I can do it now if so.

**Comment by weikengchen** [comment]
It is really to merge. I am waiting for you (technically the person who created the PR did the final decision right?)

**Pull request merged**
